### PR TITLE
feat: Adds a function to allow certificate renewal

### PR DESCRIPTION
**Pull request by saltiyazan**
# Description

This PR allows requirer charms to ask for their certificates to be renewed before those actually expire.
This was required for [this](https://github.com/canonical/vault-k8s-operator/issues/500) fix in Vault.

## Checklist

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added tests that validate the behaviour of the software.
- [x] I validated that new and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I have bumped the version of any required library.

### CODE REVIEW

File: lib/charms/tls_certificates_interface/v4/tls_certificates.py [hunk 1]
```diff
@@ -1057,12 +1057,28 @@ def _on_secret_expired(self, event: SecretExpiredEvent) -> None:
...
             csr_str = event.secret.get_content(refresh=True)["csr"]
         except ModelError:
-            logger.error("Failed to get CSR from secret - Skipping renewal")
+            logger.error("Failed to get CSR from secret - Skipping")
             return
         csr = CertificateSigningRequest.from_string(csr_str)
         self._renew_certificate_request(csr)
         event.secret.remove_all_revisions()
 
+    def renew_certificate(self, csr: str) -> None:
```
Comment (gruyaume) [inline lib/charms/tls_certificates_interface/v4/tls_certificates.py:1066]: I don't think we should pass the CSR itself here, TLS requirers don't manipulate CSR's anymore but those CertificateRequests objects that contain the CSR attributes.

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: lib/charms/tls_certificates_interface/v4/tls_certificates.py (hunk 1 above)
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v4/tls_certificates.py:1066]: Sort of related to this, on the topic of finding better names for classes, I think we should replace:
- `CertificateRequest` -> `CertificateRequestAttributes`

Comment (saltiyazan) [reply lib/charms/tls_certificates_interface/v4/tls_certificates.py]: Thanks, that's a good point, it isn't that straightforward how to get the CSR anyway even though it's part of the ProviderCertificate object that the requirer has.
I decided to use `ProviderCertificate` instead for the following reasons:
- There is no nice way to go the `CertificateRequest` --> `CertificateSigningRequest` direction and get the CSR. We'd need to do what is currently done in `_certificate_requested_for_attributes` (go the `CertificateSigningRequest` --> `CertificateRequest` direction and check if there is a match) which seems inefficient to do if there is a better way.
- The requirer already has the `ProviderCertificate` object and that object contains the `CertificateSigningRequest` which we need for the renewal.
- It feels nicer to say "_renew this certificate_" than saying "_renew the certificate that belongs to this csr_", especially that this function doesn't allow changing any attributes.

Let me know what you think.

Comment (saltiyazan) [reply lib/charms/tls_certificates_interface/v4/tls_certificates.py]: I think `CertificateRequestAttributes` is a nice name and should make things a bit clearer.
I'll go ahead and create a separate PR for that.

File: lib/charms/tls_certificates_interface/v4/tls_certificates.py [hunk 2]
```diff
@@ -1057,12 +1057,28 @@ def _on_secret_expired(self, event: SecretExpiredEvent) -> None:
...
+    def renew_certificate(self, certificate: ProviderCertificate) -> None:
+        """Request the renewal of the provided certificate."""
+        certificate_signing_request = certificate.certificate_signing_request
+        secret_label = self._get_csr_secret_label(certificate_signing_request)
+        try:
+            secret = self.model.get_secret(label=secret_label)
+            current_csr = secret.get_content(refresh=True)["csr"]
+            if current_csr != str(certificate_signing_request):
+                logger.info("No matching CSR found - Skipping")
+                return
```
Comment (gruyaume) [inline lib/charms/tls_certificates_interface/v4/tls_certificates.py:1075]: I think this block does not have to be part of the try/except loop right? only `get_secret` can trigger SecretNotFound I believe

Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v4/tls_certificates.py:1075]: Plus, in cases where we will not renew the certificate (ex. No matching CSR found or failed to get CSR from secret), we should log something higher than `info`. `warning` seems more appropriate.

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: lib/charms/tls_certificates_interface/v4/tls_certificates.py (hunk 2 above)
Comment (saltiyazan) [reply lib/charms/tls_certificates_interface/v4/tls_certificates.py]: ✅

**Review by gruyaume — APPROVED** [review]

**Pull request merged**